txpipe/rail_pz.py

```diff
@@ -0,0 +1,190 @@
+            "outpath": None,  # should not be used here
+        }
+
+        # Additional confguration specific to this algorithm
+        # can be included in the config
+        run_dict = {
+            "run_params": self.config.copy(),
+        }
+
+        # create an instance of the specific estimation
+        # method
+        cls = Estimator._find_subclass(self.config["class_name"])
+        estimator = cls(base_config, run_dict)
+
+        # Run the main training phase
+        estimator.inform()
+
+        # If there is any kind of testing/validation to be run we could
+        # do so here.
+
+        # Afterwards, save the model.  This assumes that estimator
+        # classes can be pickled, which is true for now but see the
+        # issue opened on th RAIL repo
+        with self.open_output("photoz_trained_model", wrapper=True) as output:
+            output.write(estimator)
+
+
+class PZRailEstimate(PipelineStage):
+    """Run a trained RAIL estimator to estimate PDFs and best-fit redshifts
+
+    We load a redshift Estimator model, typically saved by the PZRailTrain stage,
+    and then load chunks of photometry and run the estimator on it, and save the
+    result.
+
+    RAIL currently returns the PDF and then only the modal z as a point estimate,
+    so we save that.  Previous stages have returned mean or median methods too.
+    We could ask for this in RAIL, or calculate the mean or median ourselves.
+
+    There's currently a slight ambiguity about bin edges vs bin centers that I
+    will follow up on with the RAIL team.
+
+    The training stage is (currently all) serial, but applying the trained
+    model can be done in parallel, so we split into two stages to avoid
+    many processors sitting idle or repeating the same training process.
+
+    """
+    name = "PZRailEstimate"
+
+    inputs = [
+        ("photometry_catalog", HDFFile),
+        ("photoz_trained_model", PickleFile),
+    ]
+
+    outputs = [
+        ("photoz_pdfs", PhotozPDFFile),
+    ]
+
+    config_options = {
+        "chunk_rows": 10000,
+    }
+
+    def run(self):
+        # Importing this means that we can unpickle the relevant class
+        import rail.estimation
+
+        # Load the estimator trained in PZRailTrain
+        with self.open_input("photoz_trained_model", wrapper=True) as f:
+            estimator = f.read()
+
+        # prepare the output data - we will save things to this
+        # as we go along
+        output = self.setup_output_file(estimator)
+
+        # Create the iterator the reads chunks of photometry
+        # The method we use here automatically splits up data when we run in parallel
+        cols = [f"mag_{b}" for b in "ugrizy"] + [f"mag_err_{b}" for b in "ugrizy"]
+        chunk_rows = self.config["chunk_rows"]
+        it = self.iterate_hdf("photometry_catalog", "photometry", cols, chunk_rows)
+
+        # Loop through the chunks of data
+        for s, e, data in it:
```

Comment:
implement pooling here? especially if parallel write is available, but parallel run can always be done and gathered for sequential write ex post. Ideally all this should be hidden in the estimator codebase

Reply:
It's already automatically parallel here :-)

The `iterate_hdf` method knows about parallelism and will split up the input data among MPI processes, so the iterator only returns data chunks (including the `s` and `e` indices that are assigned to this process).  And because we opened the file with `parallel=True` in `setup_output_file` then parallel write is automatically enabled.

Reply:
why is there a for loop here?

Reply:
The `iterate_hdf` method is a generator, so it `yield`s chunks of data one by one when it's used in a for loop.

We split up both by process and by chunk, so e.g. if you have two processors and chunk size 10 with 100 data points then process 1 will be given chunks 0-9, 20-29, 40-49, 60-69, 80-89, but not all at once - it is given the one by one to minimize memory usage.

Reply:
got it.